src/hooks/with-auth.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { Component } from 'react';
+import { NextApiRequest, NextApiResponse, NextPage, NextPageContext } from 'next';
+import Router from 'next/router';
+
+import { NullableUserProfile, useUser } from './use-user';
+import { ISignInWithAuth0 } from '../instance';
+import { createLoginUrl } from '../utils/url-helpers';
+
+type RedirectToLoginProps = {
+  render: () => React.ReactElement;
+};
+
+class RedirectToLogin extends Component<RedirectToLoginProps> {
+  public constructor(props: RedirectToLoginProps) {
+    super(props);
+  }
+
+  public componentDidMount(): void {
+    window.location.assign(createLoginUrl(Router.pathname));
+  }
+
+  public render(): React.ReactElement {
+    return this.props.render();
+  }
+}
+
+type AuthenticatedProps = React.PropsWithChildren<{ user: NullableUserProfile }>;
+
+export default function withAuth(
+  InnerComponent: React.ElementType,
+  redirect: () => React.ReactElement,
+  instance: ISignInWithAuth0
```

Comment:
I wonder if it would be better for `ISignInWithAuth0` to have a `withAuth` wrapper like the handlers (rather than having to pass the instance in) eg


```js
// profile-ssr.jsx
import { withAuth } from '../lib/auth0';
```

Reply:
Good idea, so the the instance would be providing the HOC (and it would already have the instance 'closed over').